### DISCUSSION TITLE
QA wallet

### DISCUSSION THREAD

**Discussion by folex** [MetaMask/metamask-sdk#1142] [Ideas]
Hello!

While dwelling on Quality Assurance in a few web3 setups, I often wondered how to perform manual tests of dApps on real data.

Quite a few pieces fall together:
- Local chain fork, eg through `anvil`
- Account impersonation to simulate transactions from a real user's wallet, eg through `anvil_impersonateAccount`
- Balance manipulation to create test users with needed (native) token amount, eg `anvil_setBalance`

However, it's all good when sending transactions from code, or from shell. But that doesn't allow to do button-clicking in a dApp, because one piece is missing: a browser / mobile / etc wallet that can work with `anvil_impersonateAccount`, ie without actually knowing user's private key.

So basically let Metamask to import test accounts without private key and send transactions via `anvil_impersonateAccount` instead of `eth_sendTransaction`.

What do you think of such an approach? Is there any chance it is possible right now?

I think that would be invaluable for Quality Assurance.

**Comment by folex** [comment]
Seems like there's Rivet https://github.com/paradigmxyz/rivet in that niche. Haven't got it to work with my dApp yet, though.
